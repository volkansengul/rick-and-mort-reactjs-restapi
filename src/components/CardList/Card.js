import React from 'react';
import {Link} from "react-router-dom";
import Api from '../../api';


class Card extends React.Component {
    constructor(props) {
        super(props);
        this.state = {
            character: props.characterData,
            loading: false
        };
        this.api = new Api();
    }

    componentDidMount() {
        if (!this.props.characterData && this.props.characterId) {
            this.api.character(this.props.characterId).then(resp => {
                    this.setState({
                        character: resp,
                        loading: false
                    });
                    this.props.callback(resp);
                }
            );
        }
    }
    
    ready() {
        const children = this.props.children ? React.cloneElement(this.props.children, {characterData: this.state.character}) : null;
        return (
            <div className={'character-card' + (this.props.children ? ' detailed' : '')}>
                <Link to={'/detail/' + this.state.character.id}>
                    <img src={this.state.character.image}/>
                    <div>{this.state.character.name}</div>
                </Link>
                {children}
            </div>
        )
    }

    loading() {
        return <div className="loading"/>
    }

    render() {
        return this.state.character ? this.ready() : this.loading()
    }


}

class CardDetails extends React.Component {
    render() {
        return (
            <ul className={'character-card-detail'}>
                <li>
                    <span className={'caption'}> STATUS:</span>
                    <span>{this.props.characterData.status}</span>
                </li>
                <li>
                    <span className={'caption'}> SPECIES:</span>
                    <span>{this.props.characterData.species}</span>
                </li>
                <li>
                    <span className={'caption'}> GENDER:</span>
                    <span>{this.props.characterData.gender}</span>
                </li>
                <li>
                    <span className={'caption'}> ORIGIN:</span>
                    <span>{this.props.characterData.origin.name}</span>
                </li>
                <li>
                    <span className={'caption'}> LAST LOCATION:</span>
                    <span>{this.props.characterData.location.name}</span>
                </li>
            </ul>
        )
    }
}

Card.Details = CardDetails;

export default Card;