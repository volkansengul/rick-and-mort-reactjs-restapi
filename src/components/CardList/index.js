import React from 'react';
import {Redirect} from 'react-router-dom';
import Api from '../../api';
import Card from './Card';

export default class CardList extends React.Component {
    constructor() {
        super();
        this.state = {
            characters: [],
            loading: true,
            currentPage: 0,
            nextPage: undefined,
            errors: false
        };
        this.api = new Api();
    }

    nextPage() {
        this.api.characterList(this.state.nextPage).then(resp => {
                if (resp.errors !== undefined) {
                    this.setState({
                        errors: resp.errors
                    });
                } else {
                    const currentData = this.state.characters;
                    const newData = currentData.concat(resp.results);
                    this.setState({
                        characters: newData,
                        loading: false,
                        nextPage: resp.info.next
                    })
                }

            }
        );

    }

    onScroll = () => {
        if (
            (window.innerHeight + window.scrollY) >= (document.body.offsetHeight - 250)
            && this.state.loading === false && this.state.nextPage.length > 0
        ) {
            this.setState({
                loading: true
            }, this.nextPage());
        }
    };

    componentDidMount() {
        window.addEventListener('scroll', this.onScroll);
        this.nextPage();
    }

    componentWillUnmount() {
        window.removeEventListener('scroll', this.onScroll);
    }

    render() {
        const redirect = (this.state.errors !== false) ? <Redirect to="/error" push/> : null;
        const preloader = (this.state.loading) ? <div className="loading"/> : '';
        const characterCards = this.state.characters.map((item, i) => {
            return <Card key={'chracter_card_' + i} characterData={item}/>
        });
        return (
            <div className="container" ref={'container'}>
                {characterCards}
                {preloader}
                {redirect}
            </div>
        )
    }
}