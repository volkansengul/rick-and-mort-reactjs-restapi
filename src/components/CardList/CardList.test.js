import React from 'react';
import {BrowserRouter as Router} from 'react-router-dom';
import renderer from 'react-test-renderer';

import CardList from './index';
import Card from './Card';

const characterData = {
    id: 1,
    name: 'Rick Sanchez',
    image: 'https://rickandmortyapi.com/api/character/avatar/1.jpeg',
    status: 'Alive',
    gender: 'Male',
    species: 'Human',
    origin: {name: 'Earth (C-137)'},
    location: {name: 'Earth (Replacement Dimension)'}
};

it('Character Card List Test', () => {
    const card = renderer.create(
        <Router>
            <CardList characters={[characterData]}/>
        </Router>
    ).toJSON();
    expect(card).toMatchSnapshot();
});

it('Character Card Test', () => {
    const card = renderer.create(
        <Router>
            <Card characterData={characterData}/>
        </Router>
    ).toJSON();
    expect(card).toMatchSnapshot();
});


it('Character Card Detail Section Test', () => {
    const card = renderer.create(
        <Router>
            <Card.Details characterData={characterData}/>
        </Router>
    ).toJSON();
    expect(card).toMatchSnapshot();
});