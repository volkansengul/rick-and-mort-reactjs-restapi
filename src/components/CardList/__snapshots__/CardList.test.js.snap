// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Character Card Detail Section Test 1`] = `
<ul
  className="character-card-detail"
>
  <li>
    <span
      className="caption"
    >
       STATUS:
    </span>
    <span>
      Alive
    </span>
  </li>
  <li>
    <span
      className="caption"
    >
       SPECIES:
    </span>
    <span>
      Human
    </span>
  </li>
  <li>
    <span
      className="caption"
    >
       GENDER:
    </span>
    <span>
      Male
    </span>
  </li>
  <li>
    <span
      className="caption"
    >
       ORIGIN:
    </span>
    <span>
      Earth (C-137)
    </span>
  </li>
  <li>
    <span
      className="caption"
    >
       LAST LOCATION:
    </span>
    <span>
      Earth (Replacement Dimension)
    </span>
  </li>
</ul>
`;

exports[`Character Card List Test 1`] = `
<div
  className="container"
>
  <div
    className="loading"
  />
</div>
`;

exports[`Character Card Test 1`] = `
<div
  className="character-card"
>
  <a
    href="/detail/1"
    onClick={[Function]}
  >
    <img
      src="https://rickandmortyapi.com/api/character/avatar/1.jpeg"
    />
    <div>
      Rick Sanchez
    </div>
  </a>
</div>
`;
